wallet/wallet-rpc-lib/src/rpc/mod.rs

```diff
@@ -422,6 +425,47 @@ impl<N: NodeInterface + Clone + Send + Sync + 'static> WalletRpc<N> {
             .await?
     }
 
+    pub async fn sign_challenge(
+        &self,
+        account_index: U31,
+        challenge: Vec<u8>,
+        address: String,
+    ) -> WRpcResult<SignedArbitraryMessage, N> {
+        let config = ControllerConfig { in_top_x_mb: 5 }; // irrelevant
+        let destination = Address::from_str(&self.chain_config, &address)
```

Comment:
This `Address::from_str` function is a bit strange - it calls `decode_object`, ignores its result and stores the original string inside `Address`, on which you later call `decode_object` again. 
I suggest to introduce a separate function `decode_object_from_str` in Address:
```
pub fn decode_object_from_str(address: &str, cfg: &ChainConfig) -> Result<T, AddressError> {
    let data = encoding::decode(address)?;
    // here goes the rest of the body of the current "decode_object" 
}

 pub fn decode_object(&self, cfg: &ChainConfig) -> Result<T, AddressError> {
    Self::decode_object_from_str(&self.address, cfg)
}
```
and use it here and in similar places below.

Reply:
from_str needs to ensure that the encoding is valid. That's why we do it like that.